lib/pos_mempool_persister.go

```diff
@@ -0,0 +1,239 @@
+// transactions from the database based on the event type. Error is returned if the persister is not running or if there
+// is an error persisting the batch.
+func (mp *MempoolPersister) persistBatch() error {
+	if mp.status == MempoolPersisterStatusNotRunning {
+		return nil
+	}
+
+	mp.Lock()
+	defer mp.Unlock()
+
+	// If there are no transactions to persist, return.
+	if len(mp.updateBatch) == 0 {
+		return nil
+	}
+
+	err := mp.db.Update(func(txn storage.Transaction) error {
+		for _, event := range mp.updateBatch {
+			if event.Txn == nil || event.Txn.Hash == nil {
+				continue
+			}
+			// The transactions are stored in a KV database. The key is the transaction hash and the value is the
+			// serialized transaction.
+			key := event.Txn.Hash.ToBytes()
+			value, err := event.Txn.ToBytes()
+			if err != nil {
+				continue
+			}
+
+			// Set or delete a record based on the event type.
+			switch event.Type {
+			case MempoolEventAdd:
+				if err := txn.Set(key, value); err != nil {
+					glog.Errorf("MempoolPersister: Error setting key: %v", err)
+				}
+			case MempoolEventRemove:
+				if err := txn.Delete(key); err != nil {
+					glog.Errorf("MempoolPersister: Error deleting key: %v", err)
+				}
+			}
+		}
+		return nil
+	})
+	mp.updateBatch = nil
+
+	if err != nil {
+		return errors.Wrapf(err, "MempoolPersister: Error persisting batch")
+	}
+	return nil
+}
+
+// GetPersistedTransactions is used to retrieve all transactions from the database. It will return an error if the persister
+// is not currently running or if there was an issue retrieving the transactions.
+func (mp *MempoolPersister) GetPersistedTransactions() ([]*MempoolTx, error) {
+	if mp.status == MempoolPersisterStatusNotRunning {
+		return nil, errors.Wrapf(MempoolErrorNotRunning, "MempoolPersister: Cannot retrieve transactions while running")
+	}
+
+	mp.Lock()
+	defer mp.Unlock()
+
+	var mempoolTxns []*MempoolTx
+	err := mp.db.View(func(txn storage.Transaction) error {
+		// Iterate through the transaction records in the database.
+		iter, err := txn.GetIterator([]byte{})
+		if err != nil {
+			return errors.Wrapf(err, "MempoolPersister: Error retrieving iterator")
+		}
+		defer iter.Close()
+		for iter.Next() {
+			txnBytes, err := iter.Value()
+			if err != nil {
+				return errors.Wrapf(err, "MempoolPersister: Error retrieving value")
+			}
+
+			rr := bytes.NewReader(txnBytes)
+			mempoolTx := &MempoolTx{}
+			if err := mempoolTx.FromBytes(rr); err != nil {
+				return errors.Wrapf(err, "MempoolPersister: Error retrieving txn")
+			}
+			mempoolTxns = append(mempoolTxns, mempoolTx)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "MempoolPersister: Error retrieving transactions")
+	}
+	return mempoolTxns, nil
+}
+
+// OnMempoolEvent is used to add a transaction event to the eventQueue.
+func (mp *MempoolPersister) OnMempoolEvent(event *MempoolEvent) {
```

Comment:
Nit: clearer naming that decouples the persister's naming from the event based flow you have above.
```suggestion
// EnqueueEvent is used to add a transaction event to the eventQueue.
func (mp *MempoolPersister) EnqueueEvent(event *MempoolEvent) {
```